api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ValueSetTranslatorImpl.java

```diff
@@ -72,45 +70,52 @@ public ValueSet toFhirResource(@Nonnull Concept concept) {
+		sets.add(conceptUuidSet);
+		
+		Map<ConceptSource, String> conceptSourceCache = new HashMap<>();
+		
 		for (ConceptSet conceptSet : concept.getConceptSets()) {
 			Concept conceptSetMember = conceptSet.getConcept();
 			
 			if (conceptSetMember != null) {
-				if (conceptSetMember.getSet()) {
-					ValueSet.ConceptSetComponent conceptSetComponent = new ValueSet.ConceptSetComponent();
-					conceptSetComponent
-					        .addValueSet(valueSetReferenceTranslator.toFhirResource(conceptSetMember).getReference());
-					compose.addInclude(conceptSetComponent);
-					continue;
-				}
+				// first, add the uuid and display to the concept uuid set
+				ValueSet.ConceptReferenceComponent uuidConceptReferenceComponent = new ValueSet.ConceptReferenceComponent();
+				uuidConceptReferenceComponent.setCode(conceptSetMember.getUuid());
+				uuidConceptReferenceComponent.setDisplay(conceptSetMember.getDisplayString());
+				conceptUuidSet.addConcept(uuidConceptReferenceComponent);
 				
+				// now iterate through all the mappings
 				if (conceptSetMember.getConceptMappings() != null) {
-					Map<ConceptSource, String> conceptSourceCache = new HashMap<>();
-					for (ConceptMap mapping : conceptSetMember.getConceptMappings()) {
-						ValueSet.ConceptSetComponent conceptSetComponent = new ValueSet.ConceptSetComponent();
-						ValueSet.ConceptReferenceComponent conceptReferenceComponent = new ValueSet.ConceptReferenceComponent();
-						
-						ConceptReferenceTerm crt = mapping.getConceptReferenceTerm();
-						if (crt == null || crt.getConceptSource() == null) {
-							continue;
-						}
-						
+					for (ConceptMap conceptMapping : conceptSetMember.getConceptMappings()) {
+						ConceptReferenceTerm crt = conceptMapping.getConceptReferenceTerm();
 						String sourceUrl = conceptSourceCache.computeIfAbsent(crt.getConceptSource(),
 						    this::conceptSourceToURL);
-						if (sourceUrl == null) {
-							conceptReferenceComponent.setCode(conceptSetMember.getUuid());
-						} else {
-							conceptSetComponent.setSystem(sourceUrl);
+						// only add sources that we have urls for (provided by Fhir Concept Source table)
+						if (sourceUrl != null) {
+							// first see if we already have an icnlude for this source, and initialize one if not
+							ValueSet.ConceptSetComponent conceptSetComponent = sets.stream().filter(set -> {
+								return sourceUrl.equals(set.getSystem());
+							}).findFirst().orElse(null);
+							if (conceptSetComponent == null) {
+								conceptSetComponent = new ValueSet.ConceptSetComponent();
+								conceptSetComponent.setSystem(sourceUrl);
+								sets.add(conceptSetComponent);
+							}
+							// set the code and the display string
+							ValueSet.ConceptReferenceComponent conceptReferenceComponent = new ValueSet.ConceptReferenceComponent();
 							conceptReferenceComponent.setCode(crt.getCode());
+							conceptReferenceComponent.setDisplay(conceptSetMember.getDisplayString());
```

Comment:
For display values, we should use the `getMetadataTranslation()` utility I think, or a version of it. Basically, it does the same thing as the REST module's `getLocalization()` function for metadata, allowing us to use the customised values, if any.